content/blog/2019-10-23-divergence-optimizations/index.md

```diff
@@ -0,0 +1,231 @@
+
+|     Vector Instruction    | Description | Scalar Instruction |
+| ------------- | ------------- | ------------- |
+| vadd  | c[i] = a[i] + b[i]  | add |
+| vsub  | c[i] = a[i] - b[i]  | sub |
+| vmul  | c[i] = a[i] * b[i]  | mul |
+| vdiv  | c[i] = a[i] / b[i]  | div |
+| s2v  | c = (v0,v1,v2,v3)  | id |
+| s2vb  | c = (v0,v0,v0,v0)  | id |
+| v2s  | c = v[i]  | id |
+| gather  | c[0,1,2,3] = mem[a0,a1,a2,a3]  | lw |
+| scatter  | mem[a0,a1,a2,a3] = a[0,1,2,3]  | sw |
+| vload  | c[i] = mem[i]  | Can't optimize |
+| vstore  | mem[i] = a[i]  | Can't optimize |
+| vcmp  | pred = a[i] == b[i]  | Can't optimize |
+| vphi  | c[i] = pred ? a[i] : b[i]  | Can't optimize |
+
+Notably, we can't optimize across `vload` and `vstore` instructions because different memory addresses are always accessed. However, in certain cases a `gather` and `scatter` can access the exact same memory location if the address vector is the same for each. In this case, the access will be redundant, which will waste memory energy and potentially execution time. Even though we perform the `scatter`/`gather` optimization in the compiler, it is likely that the hardware implementation would also detect this case and avoid the redundant accesses.
+
+### Vector Regeneration
+
+An instruction swap could create a register type mismatch between the result of the optimized instruction and future vector instructions that use that result. For this reason, a second pass is added to the optimization algorithm. After the scalar instructions have been created, we traverse each instruction in program order and detect when a vector argument points to a scalar register. Upon detection, an `s2vb` instruction is generated to effectively cast a scalar value to a vector value. The faulting instruction argument is then updated to the new vector value produced by this instruction.
+
+The benefits of replacing a vector with a scalar outweighs the additional `s2vb` instruction. For example, a vector instruction with length four consumes three more ALU ops than a scalar instruction while an additional `s2vb` only consumes a single ALU op (we assume single-op broadcast). The overall benefit is then two ALU ops worth of energy savings.
+
+### Predication Removal
+
+Predicated vector instructions can also be simplified even in the case of a divergent predicate value. Every lane that is active in the vector instruction may still perform redundant work. Consider the Bril example below. The predicate `p0` is divergent because it's input `vec2` and `vec3` are divergent. However, the predicated vector instructions `vec4` and `vec5` are convergent because their inputs are convergent.
```

Comment:
```suggestion
Predicated vector instructions can also be simplified even in the case of a divergent predicate value. Every lane that is active in the vector instruction may still perform redundant work. Consider the Bril example below. The predicate `p0` is divergent because its inputs `vec2` and `vec3` are divergent. However, the predicated vector instructions `vec4` and `vec5` are convergent because their inputs are convergent.
```